libraries/abstractions/secure_sockets/test/iot_test_udp.c

```diff
@@ -498,6 +506,92 @@ static BaseType_t prvCheckRxTxBuffers( uint8_t * pucTxBuffer,
 }
 /*-----------------------------------------------------------*/
 
+/*
+ * @brief Test SendTo and RecvFrom echo server. Return pdFAIL if one of the operation fails, allowing retry.
+ * We expect the socket to not fail during retry.
+ */
+static BaseType_t prvSendRecvEchoTestHelper( Socket_t xSocket,
+                                             size_t xLength,
+                                             SocketsSockaddr_t * pxEchoServerAddress,
+                                             uint32_t ulTxCount )
+{
+    BaseType_t xResult;
+    BaseType_t xIntermResult;
+    uint8_t * pucTxBuffer = ( uint8_t * ) pcTxBuffer;
+    uint8_t * pucRxBuffer = ( uint8_t * ) pcRxBuffer;
+    SocketsSockaddr_t xReceiveAddress = { 0 };
+    uint32_t ulRetry;
+
+    xIntermResult = pdPASS;
+    xResult = pdFAIL;
+
+    memset( pucTxBuffer, udptestTX_BUFFER_FILLER, udptestBUFFER_SIZE );
+
+    prvCreateTxData( ( char * ) pucTxBuffer,
+                     xLength,
+                     ulTxCount );
+
+    /* Since UDP does not guarantee data delivery, we will retry until there is a
+     * successful loop */
+    for( ulRetry = 0; ulRetry < udptestMAX_RETRY + 1; ulRetry++ )
+    {
+        xIntermResult = prvSendHelper( xSocket,
+                                       pucTxBuffer,
+                                       xLength,
+                                       pxEchoServerAddress );
+
+        if( xIntermResult != pdPASS )
+        {
+            udptestPRINTF( ( "Tx %d data failed to send on try %d\r\n", ulTxCount, ulRetry ) );
+
+            /* Pause, failure may cause by network congestion */
+            vTaskDelay( 500 );
+            continue;
+        }
+
+        memset( pucRxBuffer, udptestRX_BUFFER_FILLER, udptestBUFFER_SIZE );
+        xIntermResult = prvRecvHelper( xSocket,
+                                       pucRxBuffer,
+                                       xLength,
+                                       &xReceiveAddress );
+
+        if( xIntermResult != pdPASS )
+        {
+            udptestPRINTF( ( "Rx %d data was not received on try %d\r\n", ulTxCount, ulRetry ) );
+
+            /* Pause, failure may cause by network congestion */
+            vTaskDelay( 500 );
+            continue;
+        }
+
+        xIntermResult = prvCheckRxTxBuffers( pucTxBuffer,
+                                             pucRxBuffer,
+                                             xLength );
+
+        if( xIntermResult != pdPASS )
+        {
+            udptestPRINTF( ( "Rx %d data mismatch on try %d\r\n", ulTxCount, ulRetry ) );
+
+            /* Pause, failure may cause by network congestion */
```

Comment:
Referring to earlier comment from gwicker, we need to distinguish between failure cases. On one hand, dropped and out of order packets are expected in UDP. Fragmentation is a potential issue as well. In all cases, the application code (e.g. this test) must mitigate those risks via appropriate logic. 

On the other hand, packet corruption due to the network is unlikely. If we've handled the cases above, and we're seeing corruption, that's a device bug and it should be flagged immediately rather than retried.

The net result should be that either (a) the test is able to successfully receive the transmitted buffer every time by requesting re-transmission of specific dropped data segments or (b) the test hard fails because a corrupted packet was received.

Reply:
The test now fails if send data is different from received data